### PR TITLE
Deprecate treeFromXml

### PR DESCRIPTION
**Pull request by rotu**
Clears the way for #7 in Foxy+1

### CODE REVIEW

**Review by clalancette — APPROVED** [review]
Looks good to me with green CI (don't forget to build and test `robot_state_publisher` as well).  @sloretz any thoughts here?

**Comment by rotu** [comment]
> Looks good to me with green CI

CI seems to have autorun and failed: http://build.ros2.org/job/Fpr__kdl_parser__ubuntu_focal_amd64/2/ I'm guessing because it's looking for unreleased binaries. Ran CI manually:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10609/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6049/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8639/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10504/)

> (don't forget to build and test `robot_state_publisher` as well).

`robot_state_publisher` doesn't use `treeFromXml`. Did I miss something?

**Comment by sloretz** [comment]
@rotu restarted CI testing only packages above `kdl_parser`. We're a bit tight on CI capacity right now; so please make sure to use build and test args to limit the time a job takes when possible.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10613/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6053/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8643/) <- edit Looks like mini1 needs PyQT5, will restart this job in a bit
  * rerun: [Build Status](https://ci.ros2.org/job/ci_osx/8645/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10508/)
Triggering a new build of ci_windows

**Comment by clalancette** [comment]
> > 
> 
> `robot_state_publisher` doesn't use `treeFromXml`. Did I miss something?

No, but it is the only in-tree user of kdl_parser, and it is cheap to test.  So we may as well test that things are still good there.

**Comment by rotu** [comment]
@sloretz I don't know what I'm doing! Thank you for covering for me.

**Comment by clalancette** [comment]
@ros-pull-request-builder retest this please

File: kdl_parser/include/kdl_parser/kdl_parser.hpp [hunk 1]
```diff
@@ -77,6 +77,7 @@ bool treeFromString(const std::string & xml, KDL::Tree & tree);
  * \param tree The resulting KDL Tree
  * returns true on success, false on failure
  */
+[[deprecated("Use treeFromString instead")]]
```
Comment (clalancette) [inline kdl_parser/include/kdl_parser/kdl_parser.hpp:80]: So, it turns out that we are using treeFromXml internally in https://github.com/ros2/kdl_parser/blob/ce5323ce412a679b37623ae2bcc08a074cf8410e/kdl_parser/src/kdl_parser.cpp#L161 and https://github.com/ros2/kdl_parser/blob/ce5323ce412a679b37623ae2bcc08a074cf8410e/kdl_parser/src/kdl_parser.cpp#L182 , which is triggering the deprecation warning.  For this to go in, we'll need to fix those as well to use something else.

**Review by clalancette — CHANGES_REQUESTED** [review]

File: kdl_parser/include/kdl_parser/kdl_parser.hpp (hunk 1 above)
Comment (rotu) [reply kdl_parser/include/kdl_parser/kdl_parser.hpp:80]: Addressed in #7 where the deprecation is fully resolved. I'll move those changes to this PR.

**Comment by clalancette** [comment]
@rotu So we are still getting a warning in CI about using a deprecated method: https://ci.ros2.org/job/ci_linux/10766/gcc/new/.  The problem is that `kdl_parser::initXml` is calling `urdf::Model::initXml`, which we've deprecated in ros2/urdf#12.  Two thoughts on how to fix this:

1.  Add pragmas to disable deprecation warnings when we `#include` urdf headers.  Probably a bit tricky amongst all platforms, but doable.
1.  Change the implementation of `kdl_parser::initXml` to convert the `TiXmlDocument` to a string, then call `urdf::Model::initString`.  This will hurt performance, but since our conjecture here is that nobody is using this API, that shouldn't matter too much.

I'd personally go for the second, but I'll leave it up to you how to fix it.

**Review by clalancette — APPROVED** [review]
Looking good assuming we get green CI.

**Review by jacobperron — APPROVED** [review]

**Comment by clalancette** [comment]
For future reference, CI for this was in https://github.com/ros2/rviz/pull/531#issuecomment-630808126

**Pull request merged**